File: Python/901. 股票价格跨度.py
class StockSpanner:

    def __init__(self):
        self.arr=[]
        self.dp=[]

    def next(self, price: int) -> int:
        arr=self.arr
        dp=self.dp
        if not arr:
            arr.append(price)
            dp.append(0)
            return 1
        else:
            idx=len(arr)
            if price<arr[-1]:
                dp.append(idx)
            else:
                arridx=-1
                cur=arr[arridx]
                while price>=cur:

                    arridx=dp[arridx]-1
                    if arridx==-1:
                        break
                    cur=arr[arridx]
                dp.append(arridx+1)
            arr.append(price)
            return idx-dp[-1]+1


# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)

File: Reigo/leetcode/JavaScript/901. 股票价格跨度.js
var StockSpanner = function() {
    this.idx=0;
    this.stack=[[-1,Math.pow(10,6)]];
};

/** 
 * @param {number} price
 * @return {number}
 */
StockSpanner.prototype.next = function(price) {
    var idx=this.idx;
    var stack=this.stack;
    while(stack && price>=stack[stack.length-1][1]){
        stack.pop();
    }
    stack.push([idx,price]);
    //console.log(stack)
    this.idx+=1;
    return stack[stack.length-1][0]-stack[stack.length-2][0];
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * var obj = new StockSpanner()
 * var param_1 = obj.next(price)
 */